src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -298,19 +289,25 @@ TransactionManager::submit_transaction_direct(
 {
   LOG_PREFIX(TransactionManager::submit_transaction_direct);
   SUBTRACET(seastore_t, "start", tref);
-  return trans_intr::make_interruptible(
-    tref.get_handle().enter(write_pipeline.ool_writes)
-  ).then_interruptible([this, FNAME, &tref] {
-    return seastar::do_with(tref.get_delayed_alloc_list(),
-			    [this, FNAME, &tref] (auto &delayed_extents) {
-      return epm->delayed_alloc_or_ool_write(tref, delayed_extents
-      ).si_then([this, FNAME, &tref, &delayed_extents] {
-        SUBTRACET(seastore_t, "update delayed extent mappings", tref);
-        return lba_manager->update_mappings(tref, delayed_extents);
-      }).handle_error_interruptible(
-        crimson::ct_error::input_output_error::pass_further(),
-        crimson::ct_error::assert_all("invalid error")
-      );
+  return seastar::do_with(epm->dispatch_delayed_extents(tref),
+			  [this, FNAME, &tref](auto& dispatch_res) {
+    auto usage = tref.get_projected_usage();
+    SUBDEBUGT(seastore_t, "reserve usage {}", tref, usage);
+    return trans_intr::make_interruptible(
+      epm->reserve_projected_usage(usage)
```

Comment:
Since the trim transactions need to reserve space in this PR, the real reserve order acts like a stack:
```
mutate transaction enter reserve_projected_usage phase => 
mutate transaction invokes reserve_projected_usage, and it's blocked by the trimmer, then wakeup trimmer and wait =>
  trimmer transaction does some work =>
  trimmer transaction invokes reserve_projected_usage(it also enter reserve_projected_usage phase will cause a deadlock) =>
  trimmer might be blocked by the cleaner, wakeup and wait for a temporary cleaner to get a free segment =>
    cleaner doesn't need to reserve space (when there is no cold cleaner)
    cleaner wakeup blocked trimmer =>
  trimmer finishes reservation and its work, wakeup mutate transaction =>
mutate transaction finishes reservation =>
mutate transaction leaves the reserve_projected_usage phase, which makes other transactions could enter the reserve_projected_usage phase
```

To avoid deadlock, I think only mutate transaction needs to enter the reserve_projected_usage phase and other transactions invoke `reserve_projected_usage` directly.